File: Strivers-A2Z-DSA-Sheet/14. Dynamic Programming/5. DP on Strings/python/03. Longest Common Substring.py
"""03. Longest Common Substring. Example: "ABCDGH","ACDGHR" → 4 ("CDGH")"""

def longest_common_substr(s1, s2):
    n, m = len(s1), len(s2)
    dp = [[0]*(m+1) for _ in range(n+1)]
    mx = 0
    for i in range(1, n+1):
        for j in range(1, m+1):
            if s1[i-1] == s2[j-1]:
                dp[i][j] = dp[i-1][j-1] + 1
                mx = max(mx, dp[i][j])
            else: dp[i][j] = 0
    return mx

if __name__ == "__main__":
    print(longest_common_substr("ABCDGH", "ACDGHR"))  # 4

File: Strivers-A2Z-DSA-Sheet/14. Dynamic Programming/5. DP on Strings/javascript/03. Longest Common Substring.js
/*
=============================================================================
  QUESTION: Longest Common Substring (GFG)
=============================================================================

  Find length of longest common substring (must be contiguous).
  Example: "ABCDGH", "ACDGHR" → 4 ("CDGH")

  Key diff from LCS: if chars don't match → dp[i][j] = 0 (reset).
  Track global max.

=============================================================================
  APPROACH: Tabulation — O(N×M) Time, O(N×M) Space
=============================================================================
*/

function longestCommonSubstr(s1, s2) {
    let n = s1.length, m = s2.length;
    let dp = Array.from({ length: n + 1 }, () => new Array(m + 1).fill(0));
    let maxLen = 0;
    for (let i = 1; i <= n; i++) {
        for (let j = 1; j <= m; j++) {
            if (s1[i - 1] === s2[j - 1]) {
                dp[i][j] = dp[i - 1][j - 1] + 1;
                maxLen = Math.max(maxLen, dp[i][j]);
            } else {
                dp[i][j] = 0;  // reset — must be contiguous
            }
        }
    }
    return maxLen;
}

// ==========================================================================
console.log(longestCommonSubstr("ABCDGH", "ACDGHR"));  // 4
console.log(longestCommonSubstr("abc", "abc"));          // 3

